src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,162 @@
+
+        Ok(self)
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `SafePathBuf::root()`, otherwise error will be returned.
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<SafePathBuf> {
+        let mut root = self.root.canonicalize()?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        let path = safe_join("/", path)?;
+        let mut suffix = path.strip_prefix(&root).map_err(|_| {
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", path.display()),
+            )
+        })?;
+        if suffix.file_name().is_none() {
+            return SafePathBuf::from_path(root);
+        }
+        if !self.recursive {
+            if let Some(parent) = path.parent() {
+                root = root.join(parent);
+            }
+            // Safe to unwrap() because we have verified `suffix` is not empty.
+            suffix = Path::new(suffix.file_name().unwrap());
+        }
+
+        for comp in suffix {
+            let file = SafePathBuf::from_path(&root)?;
+            if !file.target().is_dir() {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("Invalid path: {}", root.display()),
+                ));
+            }
+            root = root.join(comp);
+            DirBuilder::new()
+                .mode(self.mode)
+                .recursive(true)
+                .create(&root)?;
+        }
+
+        let result = SafePathBuf::from_path(&root)?;
+        if !result.target().is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        Ok(result)
+    }
+}
+
+impl Default for SafeDirBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+
+    #[test]
+    fn test_safe_path_buf() {
+        let rootfs_dir = tempfile::tempdir().expect("failed to create tmpdir");
+        let rootfs_path = rootfs_dir.path();
+        let mut builder = SafeDirBuilder::new();
+
+        builder
+            .root(rootfs_path.join("__does_not_exist__"))
+            .unwrap_err();
+
+        fs::write(rootfs_path.join("txt"), "test").unwrap();
+        builder.root(rootfs_path.join("txt")).unwrap_err();
+
+        builder.root(rootfs_path).unwrap();
+        builder.create("/txt/a").unwrap_err();
+        let path = builder.create(rootfs_path.join(".")).unwrap();
+        assert_eq!(path.target(), rootfs_path);
+        builder.create(rootfs_path.join("a/b/c")).unwrap_err();
```

Comment:
You should test `a/b` here as well, since that's the *minimal* case that shouldn't work with `recursive` set to false.

Reply:
fixed